scheduler/src/cook/mesos/util.clj

```diff
@@ -362,51 +395,55 @@
-                                 (= (.v %) user)))
-               (map #(.e %))
+          (->> (d/seek-datoms db :avet :job/user user entid-start)
+               (take-while #(and (< (:e %) entid-end)
+                                 (= (:a %) job-user-entid)
+                                 (= (:v %) user)))
+               (map #(:e %))
                (map (partial d/entity db))
-               (filter #(<= (.getTime start) (.getTime (:job/submit-time %))))
-               (filter #(< (.getTime (:job/submit-time %)) (.getTime end)))
+               (filter #(let [^long submit-ms (.getTime (:job/submit-time %))]
+                          (and (<= start-ms submit-ms)
+                               (< submit-ms end-ms))))
                (filter #(= :job.state/completed (:job/state %)))
                (filter #(not (:job/custom-executor %))))]
       (->>
         (cond->> jobs
                  (instance-states state) (filter #(= state (job-ent->state %)))
                  name-filter-fn (filter #(name-filter-fn (:job/name %))))
-        (take limit)))))
+        (take limit)
+        doall))))
 
-;; This is a separate query from completed jobs because most running/waiting jobs
-;; will be at the end of the time range, so the query is somewhat inefficent.
-;; The standard datomic query performs reasonably well on the smaller set of
-;; running and waiting jobs, so it's implemented that way to keep things simple.
 (defn get-active-jobs-by-user-and-state
   "Returns all jobs for a particular user in the specified state
-   and timeframe, without a custom executor.
-   Note that this query is not performant for completed jobs, use
-   get-completed-jobs-by-user instead."
-  [db user start end state name-filter-fn]
+   and timeframe, without a custom executor. This query looks for all
+   jobs by job state only (i.e., no 'success' or 'failed')"
+  [db user start end state limit name-filter-fn]
   (let [state-keyword (case state
                         "running" :job.state/running
                         "waiting" :job.state/waiting)
-        jobs
-        (timers/time!
-          (timers/timer ["cook-mesos" "scheduler" (str "get-" (name state) "-jobs-by-user-duration")])
-          (->> (q '[:find [?j ...]
-                    :in $ ?user ?state ?start ?end
-                    :where
-                    [?j :job/state ?state]
-                    [?j :job/user ?user]
-                    [?j :job/submit-time ?t]
-                    [(<= ?start ?t)]
-                    [(< ?t ?end)]
-                    [?j :job/custom-executor false]]
-                  db user state-keyword start end)
-               (map (partial d/entity db))))]
-    (cond->> jobs name-filter-fn (filter #(name-filter-fn (:job/name %))))))
+        millis-diff (max 0 (- (.getTime end) (.getTime start) 60000)) ; 60000ms is so that we get queries that are just a tiny bit over a bucket.
+        hours-diff (/ millis-diff 3600000)
```

Comment:
`Please use the `(-> 1 t/hours t/in-millis)` so it is easier to understand what time magnitude we are using

Reply:
ACK